DS-3636 Indexes and caching performance updates

jonas-atmire opened the pull request:
Pull Request for JIRA ticket:
https://jira.duraspace.org/browse/DS-3636

jonas-atmire committed:
DS-3636 Indexes and caching performance updates

jonas-atmire committed:
DS-3636 Use resourcePolicyService to create and update the copied policy

tomdesair commented on dspace-api/src/main/java/org/dspace/authorize/AuthorizeServiceImpl.java:
```diff
@@ -612,21 +612,10 @@ public void addPolicies(Context c, List<ResourcePolicy> policies, DSpaceObject d
-            // and add policy to list of new policies
-            newPolicies.add(rp);
+            ResourcePolicy policy = resourcePolicyService.create(c);
+            policy = srp.copyToDSpaceObject(policy,dest);
```
This kind of beats the purpose of having the "copyToDSpaceObject" method. This code will force hibernate to execute two DB statements: an insert and an update. 

I think you should move the copyToDSpaceObject to the ResourcePolicyService:
* Create a new empty resource policy
* Copy over the values
* Then pass that filled in resource policy to `resourcePolicyDAO.create()`.

You can enable Hibernate DEBUG logging in hibernate.cfg.xml to see the SQL statements it executes.

tomdesair commented on dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/h2/V6.0_2017.07.04__DS-3636_General_Index_Performance_Updates.sql:
```diff
@@ -0,0 +1,32 @@
+
+CREATE INDEX resourcepolicy_action_idx ON resourcepolicy (action_id);
+CREATE INDEX resourcepolicy_resource_type_id_idx ON resourcepolicy(resource_type_id );
+CREATE INDEX resourcepolicy_resource_id_idx ON resourcepolicy(resource_id );
```
This column does not exist in DSpace 6! Please re-test.

tomdesair commented on dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/h2/V6.0_2017.07.04__DS-3636_General_Index_Performance_Updates.sql:
```diff
@@ -0,0 +1,32 @@
+CREATE INDEX group2groupcache_parent_id_idx ON group2groupcache (parent_id);
+CREATE INDEX group2groupcache_child_id_idx ON group2groupcache (child_id);
+CREATE INDEX bundle2bitstream_bitstream_order_idx ON bundle2bitstream (bitstream_order);
+CREATE INDEX metadatavalue_mf_place_idx ON metadatavalue (metadata_field_id, place);
```
There is already an index on metadata_field_id, so maybe we should just index "place".

tomdesair commented on dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/h2/V6.0_2017.07.04__DS-3636_General_Index_Performance_Updates.sql:
```diff
@@ -0,0 +1,32 @@
+CREATE INDEX versionitem_item_id_idx ON versionitem (item_id);
+CREATE INDEX versionitem_versionhistory_id_idx ON versionitem (versionhistory_id);
+CREATE INDEX group2groupcache_parent_id_idx ON group2groupcache (parent_id);
+CREATE INDEX group2groupcache_child_id_idx ON group2groupcache (child_id);
```
The above two indexes already exist in DSpace 6: https://github.com/DSpace/DSpace/blob/dspace-6.0/dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/postgres/V6.0_2015.03.07__DS-2701_Hibernate_migration.sql#L206

tomdesair reviewed (changes requested):
There are still some issues with this port to DSpace 6.

Commit:
DS-3563: Fix Oracle Flyway migration error

jonas-atmire committed:
DS-3636 Oracle flyway fixes + additional PR feedback

jonas-atmire commented:
Updated the PR based on the feedback:
- Removed the defining of pre-existing indexes
- Updated "resourcepolicy_resource_id_idx" index to use the "dspace_object" column instead
- Updated the metadatavalue_mf_place_idx to only index "place"
- Moved the "copyToDSpaceObject" to the service -> This now only uses a single statement, which creates the entire dspaceobject

jonas-atmire committed:
DS-3636 Removal of duplicate indexing with different name

jonas-atmire commented:
The index on the dspace object was already present in the following flyway script, removed it again
https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/postgres/V6.0_2015.03.07__DS-2701_Hibernate_migration.sql#L341

Checked that there were no more duplicate indexes with different names applied on the tables.

Commit:
DS-3636: Remove duplicate index creation

AlexanderS commented:
Tested this and the flyway migrations are executed with the following messages:

```
2017-07-06 16:31:55,798 INFO  org.flywaydb.core.internal.command.DbValidate @ Successfully validated 33 migrations (execution time 00:00.031s)
2017-07-06 16:31:55,845 INFO  org.flywaydb.core.internal.command.DbMigrate @ Current version of schema "public": 6.0.2016.08.23
2017-07-06 16:31:55,846 INFO  org.flywaydb.core.internal.command.DbMigrate @ Migrating schema "public" to version 6.0.2017.07.04 - DS-3636 General Index Performance Updates
2017-07-06 16:31:55,866 INFO  org.flywaydb.core.internal.dbsupport.JdbcTemplate @ DB: index "resourcepolicy_action_idx" does not exist, skipping
2017-07-06 16:31:55,867 INFO  org.flywaydb.core.internal.dbsupport.JdbcTemplate @ DB: index "resourcepolicy_resource_type_id_idx" does not exist, skipping
2017-07-06 16:31:55,867 INFO  org.flywaydb.core.internal.dbsupport.JdbcTemplate @ DB: index "resourcepolicy_resource_id_idx" does not exist, skipping
2017-07-06 16:31:55,868 INFO  org.flywaydb.core.internal.dbsupport.JdbcTemplate @ DB: index "metadatavalue_mf_place_idx" does not exist, skipping
2017-07-06 16:31:55,868 INFO  org.flywaydb.core.internal.dbsupport.JdbcTemplate @ DB: index "bundle2bitstream_bitstream_order_idx" does not exist, skipping
2017-07-06 16:31:55,914 INFO  org.flywaydb.core.internal.command.DbMigrate @ Successfully applied 1 migration to schema "public" (execution time 00:00.086s).
```

So should be all fine (I only have postgres so I cannot test the oracle stuff). Tested all the basic stuff (browsing, search, download and submission) and all seems to work fine.

Then I removed the last migration from the `schema_version` table and restarted the server:

```
2017-07-06 16:40:46,714 INFO  org.flywaydb.core.internal.command.DbValidate @ Successfully validated 33 migrations (execution time 00:00.027s)
2017-07-06 16:40:46,783 INFO  org.flywaydb.core.internal.command.DbMigrate @ Current version of schema "public": 6.0.2016.08.23
2017-07-06 16:40:46,783 INFO  org.flywaydb.core.internal.command.DbMigrate @ Migrating schema "public" to version 6.0.2017.07.04 - DS-3636 General Index Performance Updates
2017-07-06 16:40:46,801 INFO  org.flywaydb.core.internal.dbsupport.JdbcTemplate @ DB: index "resourcepolicy_resource_id_idx" does not exist, skipping
2017-07-06 16:40:46,808 ERROR org.flywaydb.core.internal.command.DbMigrate @ Migration of schema "public" to version 6.0.2017.07.04 - DS-3636 General Index Performance Updates failed! Changes successfully rolled back.
2017-07-06 16:40:46,810 FATAL org.dspace.core.Context @ Cannot initialize database via Flyway!
java.sql.SQLException: Flyway migration error occurred
        at org.dspace.storage.rdbms.DatabaseUtils.updateDatabase(DatabaseUtils.java:666)
        at org.dspace.storage.rdbms.DatabaseUtils.updateDatabase(DatabaseUtils.java:574)
        at org.dspace.storage.rdbms.DatabaseUtils.updateDatabase(DatabaseUtils.java:550)
        at org.dspace.core.Context.<clinit>(Context.java:103)
        at org.dspace.app.util.AbstractDSpaceWebapp.register(AbstractDSpaceWebapp.java:74)
        at org.dspace.app.util.DSpaceWebappListener.contextInitialized(DSpaceWebappListener.java:31)
        at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:5027)
        at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5525)
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
        at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:901)
        at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:877)
        at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:649)
        at org.apache.catalina.startup.HostConfig.deployDescriptor(HostConfig.java:672)
        at org.apache.catalina.startup.HostConfig$DeployDescriptor.run(HostConfig.java:1861)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.flywaydb.core.internal.dbsupport.FlywaySqlScriptException:
Migration V6.0_2017.07.04__DS-3636_General_Index_Performance_Updates.sql failed
-------------------------------------------------------------------------------
SQL State  : 42P07
Error Code : 0
Message    : ERROR: relation "versionitem_versionhistory_id_idx" already exists
Location   : org/dspace/storage/rdbms/sqlmigration/postgres/V6.0_2017.07.04__DS-3636_General_Index_Performance_Updates.sql (/var/lib/tomcat7/file:/opt/dspace/webapps/xmlui/WEB-INF/lib/dspace-api-6.1-SNAPSHOT.jar!/org/dspace/storage/rdbms/sqlmigration/postgres/V6.0_2017.07.04__DS-3636_General_Index_Performance_Updates.sql)
Line       : 22
Statement  : CREATE INDEX versionitem_versionhistory_id_idx ON versionitem (versionhistory_id)

        at org.flywaydb.core.internal.dbsupport.SqlScript.execute(SqlScript.java:117)
        at org.flywaydb.core.internal.resolver.sql.SqlMigrationExecutor.execute(SqlMigrationExecutor.java:71)
        at org.flywaydb.core.internal.command.DbMigrate.doMigrate(DbMigrate.java:352)
        at org.flywaydb.core.internal.command.DbMigrate.access$1100(DbMigrate.java:47)
        at org.flywaydb.core.internal.command.DbMigrate$4.doInTransaction(DbMigrate.java:308)
        at org.flywaydb.core.internal.util.jdbc.TransactionTemplate.execute(TransactionTemplate.java:72)
        at org.flywaydb.core.internal.command.DbMigrate.applyMigration(DbMigrate.java:305)
        at org.flywaydb.core.internal.command.DbMigrate.access$1000(DbMigrate.java:47)
        at org.flywaydb.core.internal.command.DbMigrate$2.doInTransaction(DbMigrate.java:230)
        at org.flywaydb.core.internal.command.DbMigrate$2.doInTransaction(DbMigrate.java:173)
        at org.flywaydb.core.internal.util.jdbc.TransactionTemplate.execute(TransactionTemplate.java:72)
        at org.flywaydb.core.internal.command.DbMigrate.migrate(DbMigrate.java:173)
        at org.flywaydb.core.Flyway$1.execute(Flyway.java:959)
        at org.flywaydb.core.Flyway$1.execute(Flyway.java:917)
        at org.flywaydb.core.Flyway.execute(Flyway.java:1373)
        at org.flywaydb.core.Flyway.migrate(Flyway.java:917)
        at org.dspace.storage.rdbms.DatabaseUtils.updateDatabase(DatabaseUtils.java:655)
        ... 18 more
Caused by: org.postgresql.util.PSQLException: ERROR: relation "versionitem_versionhistory_id_idx" already exists
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2458)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2158)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:291)
        at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:432)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:358)
        at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:305)
        at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:291)
        at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:269)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:265)
        at org.apache.commons.dbcp2.DelegatingStatement.execute(DelegatingStatement.java:291)
        at org.apache.commons.dbcp2.DelegatingStatement.execute(DelegatingStatement.java:291)
        at org.flywaydb.core.internal.dbsupport.JdbcTemplate.executeStatement(JdbcTemplate.java:238)
        at org.flywaydb.core.internal.dbsupport.SqlScript.execute(SqlScript.java:114)
        ... 34 more
```

The index name of the third `DROP INDEX IF EXISTS` does not match with the name of the third `CREATE INDEX`.

jonas-atmire committed:
DS-3636 Flyway index ordering + comments update

jonas-atmire commented:
@AlexanderS 
Good catch, it seems that for both the oracle and h2 flyway files, the line that initially drops the "versionitem_versionhistory_id_idx" index was missing.
This had previously been accidentally removed during the removal of actually non-required indexes.

I've updated:
- The comment to represent/mention only the used indexes and tables.
- The order of the h2 and postgres index dropping and creation to avoid confusion
- The naming of the third index, which was actually the one that needed removing in the first place.

Retested this for both postgres and oracle by removing the schema_version row containing this update and rerunning flyway. (Haven't tested the h2 flyway, but this is the same file as for postgres, so I don't think this is a problem?)

AlexanderS reviewed (approved):
Yes, it works now without the issue for Postgres.

terrywbrady reviewed (changes requested):
Please add more context for this PR to help explain its purpose.  Based on a quick review, I have concerns about some of the proposed indexes.

terrywbrady commented on dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/h2/V6.0_2017.07.04__DS-3636_General_Index_Performance_Updates.sql:
```diff
@@ -0,0 +1,24 @@
+CREATE INDEX resourcepolicy_resource_type_id_idx ON resourcepolicy(resource_type_id );
+CREATE INDEX versionitem_versionhistory_id_idx ON versionitem (versionhistory_id);
+CREATE INDEX bundle2bitstream_bitstream_order_idx ON bundle2bitstream (bitstream_order);
+CREATE INDEX metadatavalue_mf_place_idx ON metadatavalue (place);
```
This field should not be indexed on its own.  There would never be a need to retrieve items by place alone.

terrywbrady commented on dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/h2/V6.0_2017.07.04__DS-3636_General_Index_Performance_Updates.sql:
```diff
@@ -0,0 +1,24 @@
+CREATE INDEX resourcepolicy_action_idx ON resourcepolicy (action_id);
+CREATE INDEX resourcepolicy_resource_type_id_idx ON resourcepolicy(resource_type_id );
+CREATE INDEX versionitem_versionhistory_id_idx ON versionitem (versionhistory_id);
+CREATE INDEX bundle2bitstream_bitstream_order_idx ON bundle2bitstream (bitstream_order);
```
This field should not be indexed on its own.  There would never be a need to retrieve items by bitstream_order alone.

terrywbrady commented on dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/h2/V6.0_2017.07.04__DS-3636_General_Index_Performance_Updates.sql:
```diff
@@ -0,0 +1,24 @@
+DROP INDEX IF EXISTS bundle2bitstream_bitstream_order_idx;
+DROP INDEX IF EXISTS metadatavalue_mf_place_idx;
+
+CREATE INDEX resourcepolicy_action_idx ON resourcepolicy (action_id);
```
This field should not be indexed on its own.  There would never be a need to retrieve items by action_id alone.

terrywbrady commented on dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/h2/V6.0_2017.07.04__DS-3636_General_Index_Performance_Updates.sql:
```diff
@@ -0,0 +1,24 @@
+DROP INDEX IF EXISTS metadatavalue_mf_place_idx;
+
+CREATE INDEX resourcepolicy_action_idx ON resourcepolicy (action_id);
+CREATE INDEX resourcepolicy_resource_type_id_idx ON resourcepolicy(resource_type_id );
```
This field should not be indexed on its own.  There would never be a need to retrieve items by resource_type_id alone.

terrywbrady commented on dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/postgres/V6.0_2017.07.04__DS-3636_General_Index_Performance_Updates.sql:
```diff
@@ -0,0 +1,24 @@
+DROP INDEX IF EXISTS versionitem_versionhistory_id_idx;
+DROP INDEX IF EXISTS bundle2bitstream_bitstream_order_idx;
+DROP INDEX IF EXISTS metadatavalue_mf_place_idx;
+
```
See other comments about these indexes.

terrywbrady reviewed (dismissed):
The proposed indexes would be very inefficient if created.  Also, a detailed summary is needed to explain the purpose of this PR.  If it is no longer applicable, please close it.

tomdesair commented:
Hi @terrywbrady,

These indexes are not inefficient as RDBMS's are able to combine different indexes as if they were "multi-column" indexes. So having single-column indexes separately increase the probability that the index can be reused when writing other (custom) queries (for custom code). 

More information on how Postgres is able to combine indexes can be found here: https://www.postgresql.org/docs/9.6/static/indexes-bitmap-scans.html

terrywbrady commented:
@AlexanderS , @tomdesair , I am sorry that I missed Tom's comment on this ticket from last Sept.  I had forgotten about this PR.  It seems like it would be good to get another reviewer to look at this PR.  

I flagged this as a concern based on my past understanding of database indexes.  Perhaps some comments could be added around the index definitions to explain these changes.  It would also be useful for you to describe in the PR the testing that has been done to confirm the performance improvement.  

I also recommend flagging this PR for discussion at the next DEV meeting.

mwoodiupui commented:
See also for Oracle:  https://jonathanlewis.wordpress.com/2007/02/08/index-combine/
That is a little deep for me, but it seems to be talking about something similar to the PostgreSQL article linked above.

mwoodiupui commented on dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/oracle/V5.7_2017.04.11__DS-3563_Index_metadatavalue_resource_type_id_column.sql:
```diff
@@ -19,5 +19,5 @@ begin
   when index_not_exists then null;
 end;
-
+/
```
Note for the Oracle-impaired (like me):  this slash is needed to cause execution of the block above, because the block swallows the semicolon on line 21.  See https://stackoverflow.com/questions/1079949/when-do-i-need-to-use-a-semicolon-vs-a-slash-in-oracle-sql

mwoodiupui reviewed (approved):
+1 by inspection.  This all looks reasonable.

tdonohue commented:
I'm unclear if this has received recent testing? (against 6.3)  I just realized this PR looks to be at +2.  

The changes look OK to me overall, but as the commits are from 2017 and this involves database migrations, it might be worth a quick "rebase" to ensure everything still works fine on latest `dspace-6_x`.
